source/usage-examples/convenient-transactions.txt

```diff
@@ -0,0 +1,162 @@
+
+.. important:: Return Types and Values
+   
+   Starting in version 6.0 of the {+driver-short+}, these methods return
+   the value that the provided function returns.
+
+   For example, if a function you provide to the ``withTransaction()``
+   method returns a ``Document`` type, then the ``withTransaction()``
+   method returns that object. Furthermore, if you perform this
+   transaction within the function you provide to the ``withSession()``
+   method, then the ``withSession()`` method also returns the
+   ``Document`` object.
+
+   If you structure your code in this way, you can propagate return
+   values out of the session and work with them elsewhere in your code.
+
+Example
+-------
+
+Consider a situation where a customer purchases items from your store.
+Whenever a customer attempts to purchase an item, your application must
+check if the items are available, and if so, record details about the
+purchase.
+
+You record the inventory in the ``inventory`` collection of the
+``testdb`` database. Before any transactions, this collection contains
+the following documents:
+
+.. code-block:: json
+
+   { item: "sunblock, 8 oz", qty: 85, price: 6.0 },
+   { item: "beach chair", qty: 30, price: 25.0 },
+
+You store purchase records in the ``records`` collection of the
+``testdb`` database. This collection is empty, as there have been no
+purchases.
+
+This example code performs a transaction through the following actions:
+
+1. Calls the ``withSession()`` method on the client to implicitly create
+   the session and include the transaction within this session.
+
+2. Calls the ``withTransaction()`` method on the session to create a
+   transaction, run the specified operations, and commit the
+   transaction. If the transaction fails, this method aborts the
+   transaction and returns an error message.
+
+3. Performs the following operations within the transaction:
+
+- Processing a predefined ``order`` variable that specifies the item and
+  quantity that the customer wants, the first operation checks if
+  sufficient inventory is available to fulfill the purchase.
+- If the items are not available or there are not enough items, the
+  transaction aborts and the function returns an error message.
+- If the inventory is sufficient for the order, the code performs an
+  update operation in the ``inventory`` collection to reduce the
+  quantities of the items purchased.
+- The code inserts a record of the purchase to the ``records``
+  collection that includes a timestamp, a list of the purchased
+  items, and a total price for the purchase.
+- The code returns a message acknowledging that the transaction
+  committed successfully with a copy of the purchase record.
+
+4. Prints the return type of ``withSession()``, which is either the
+   error message or the acknowledgment that the transaction completed.
+
+.. note::
+   
+   You must pass the session instance to each operation that
+   you want to run in that session.
```

Comment:
Probably good adds here:
- You must also `await` every operation 
  - we've seen this missed in JIRA tickets in the past
- You must NOT use parallelism (`Promise.all`)
  - Sessions cannot be used in parallel, but also are unable to report an error if they are used this way. It will likely lead to a server error, but inconsistently depending on timing.